Mathlib/RingTheory/MatrixAlgebra.lean

```diff
@@ -92,133 +92,140 @@ variable [CommSemiring R]
-      simp_rw [Matrix.mul_apply, Matrix.smul_apply, Matrix.map_apply, smul_eq_mul, Finset.mul_sum,
-        _root_.mul_assoc, Algebra.left_comm])
-    (by
-      simp_rw [toFunLinear, lift.tmul, toFunBilinear_apply,
-        Matrix.map_one (algebraMap R A) (map_zero _) (map_one _), one_smul])
+abbrev toTensorMatrixLin : K ⊗[F] Matrix n n A →ₗ[K] Matrix n n (K ⊗[F] A) :=
+  AlgebraTensorModule.lift
+    { toFun k := (TensorProduct.mk F _ _ k).mapMatrix
+      map_add' k1 k2 := by ext; simp [add_smul]
+      map_smul' r k := by ext; dsimp; rw [← smul_eq_mul, smul_tmul'] }
 
 @[simp]
-theorem toFunAlgHom_apply (a : A) (m : Matrix n n R) :
-    toFunAlgHom n R A (a ⊗ₜ m) = a • m.map (algebraMap R A) := rfl
+theorem toTensorMatrixLin_tmul (k : K) (M : Matrix n n A) :
+    toTensorMatrixLin K F A n (k ⊗ₜ M) = M.map (fun x => k ⊗ₜ x) := rfl
 
-/-- (Implementation detail.)
+variable [DecidableEq n]
+
+@[simp high]
+theorem toTensorMatrixLin_tmul_single (k : K) (i : n) (j : n) (a : A) :
+    toTensorMatrixLin K F A n (k ⊗ₜ .single i j a) = .single i j (k ⊗ₜ a) := by
+  simp_rw [toTensorMatrixLin_tmul, ← mk_apply, map_single]
 
-The bare function `Matrix n n A → A ⊗[R] Matrix n n R`.
-(We don't need to show that it's an algebra map, thankfully --- just that it's an inverse.)
--/
-def invFun (M : Matrix n n A) : A ⊗[R] Matrix n n R :=
-  ∑ p : n × n, M p.1 p.2 ⊗ₜ single p.1 p.2 1
+
+private def invFun_toFun_bilinear (i j : n): K →ₗ[F] A →ₗ[F] K ⊗[F] Matrix n n A :=
+  AlgebraTensorModule.mk F F K (Matrix n n A) |>.compl₂ (Matrix.singleLinearMap _ i j)
 
 @[simp]
-theorem invFun_zero : invFun n R A 0 = 0 := by simp [invFun]
+lemma invFun_toFun_bilinear_apply (i j : n) (k : K) (a : A) :
+    invFun_toFun_bilinear K F A n i j k a = k ⊗ₜ single i j a := rfl
+
+private abbrev invFun_toFun (i j : n) : K ⊗[F] A →ₗ[F] K ⊗[F] Matrix n n A :=
+  TensorProduct.lift <| invFun_toFun_bilinear K F A n i j
+
+private abbrev invFun_Klinear (i j : n) : K ⊗[F] A →ₗ[K] K ⊗[F] Matrix n n A where
+  __ := invFun_toFun K F A n i j
+  map_smul' k tensor := by
+    induction tensor with
+    | zero => simp
+    | tmul k0 a => simp [smul_tmul', MulAction.mul_smul]
+    | add _ _ h1 h2 => simp_all
+
+variable [Fintype n]
+
+private def invFun_linearMap : Matrix n n (K ⊗[F] A) →ₗ[K] K ⊗[F] Matrix n n A :=
+  (LinearMap.lsum K _ ℕ fun ii => LinearMap.lsum K _ ℕ fun jj =>
+    invFun_Klinear K F A n ii jj) ∘ₗ (ofLinearEquiv K).symm.toLinearMap
 
 @[simp]
-theorem invFun_add (M N : Matrix n n A) :
-    invFun n R A (M + N) = invFun n R A M + invFun n R A N := by
-  simp [invFun, add_tmul, Finset.sum_add_distrib]
+theorem invFun_linearMap_single_tmul (i j : n) (k : K) (a : A) :
+    invFun_linearMap K F A n (.single i j (k ⊗ₜ a)) = k ⊗ₜ .single i j a := by
+  simp [invFun_linearMap, -LinearMap.lsum_apply, LinearMap.lsum_piSingle]
+
+attribute [local ext] AlgebraTensorModule.ext Matrix.ext_linearMap in
+def linearEquivTensor : K ⊗[F] Matrix n n A ≃ₗ[K] Matrix n n (K ⊗[F] A) :=
```

Comment:
This should generalize to rectangular matrices over modules, and also a tower of three rings rather than two.